Mathlib/Data/Set/Subset.lean

```diff
@@ -0,0 +1,260 @@
+/-
+Copyright (c) 2024 Miguel Marco. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Miguel Marco
+-/
+import Mathlib.Data.Set.Function
+import Mathlib.Data.Set.Functor
+import Mathlib.Lean.Expr.ExtraRecognizers
+
+/-!
+# Sets in subtypes
+
+This file defines notation for sets in a type pulled down to sets in a subtype, and sets in
+a subtype, lifted up to sets in the ambient type.
+
+It also provides some related lemmas for convenience.
+
+## Notation
+
+Let `α` be a `Type`, `A B : Set α` two sets in `α`, and `C : Set ↑A` a set in the subtype `↑A`.
+
+- `A ↓∩ B` denotes `(Subtype.val ⁻¹' B : Set A)` (that is, `{x : ↑A | ↑x ∈ B}`).
+- `↑C` denotes `Subtype.val '' C` (that is, `{x : α | ∃ y ∈ C, ↑y = x}`).
+
+To access this notation, type `open Subset`.
+
+## Tags
+
+subsets
+-/
+
+open Set
+
+universe u
+
+variable {α β : Type u} {A B C: Set α} {D E : Set ↑A}
+variable {S : Set (Set α)} {T : Set (Set ↑A)} {i : β → Set α} {j : β → Set ↑A}
+
+namespace Subset
+
+/--
+Given two sets `A` and `B`, `A ↓∩ B` denotes `{x : ↑A | ↑x ∈ B}`.
+-/
+scoped notation3 A:67 " ↓∩ " B:67 => (Subtype.val ⁻¹' (B : type_of% A) : Set (A : Set _))
+
+open Lean PrettyPrinter Delaborator SubExpr in
+/--
+Sets of a subtype coerced to the ambient type are denoted with `↑`.
+-/
+@[scoped delab app.Set.image]
+def delab_set_image_subtype : Delab := do
+  let #[α, _, f, _] := (← getExpr).getAppArgs | failure
+  guard <| f.isAppOfArity ``Subtype.val 2
+  let some _ := α.coeTypeSet? | failure
+  let e ← withAppArg delab
+  `(↑$e)
```

Comment:
This delaborator should go right after the `CoeHead (Set s) (Set α)` instance in `Mathlib/Data/Set/Functor.lean`. I don't think it needs to be `scoped`, given that the coercion instance isn't scoped. (Make sure to move the `Mathlib.Lean.Expr.ExtraRecognizers` import.)

Could you update the docstring to mention that this is the delaborator for the `CoeHead (Set s) (Set α)` instance?

Reply:
Eric Wieser [commented ](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Should.20Sets.20be.20coerced.3F/near/417208299) that not having this delaborator working by default could be a feature, since it could help realising what kind of lemmas one should search for.

That is why I decided to keep this delaborator scoped.

Reply:
Ok, that's fine. I think we should still move it to be close to the `CoeHead` instance. The module doc in this file can still mention the notation and `Mathlib.Data.Set.Functor` for discoverability.

Reply:
I moved the notation to `Mathlib.Data.Set.Functor`, but I am not sure if the documentation is clear enough now.